File: Java/Leetcode19.java
package Java;

public class Leetcode19 {
    class ListNode {
       int val;
       ListNode next;
       ListNode() {}
       ListNode(int val) { this.val = val; }
       ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    // Sliding window technique
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode start = head;
        ListNode end = head;

        // i~j : size N window
        while(n>0 && end!= null){
            end = end.next;
            n -= 1;
        }

        // if there is only one element, remove that one
        if(end==null){head = head.next;}
        else{
            // window reaches the end
            while(end.next != null){
                start = start.next;
                end = end.next;
            }
            // remove the element at the start
            start.next = start.next.next;
        }
        return head;
    }
}

// if n=3
// 1 2 3 4 5
// <--->
// move the window until end
// 1 2 3 4 5
//     <--->
// remove the start

File: Python/Leetcode19.py
class Leetcode19:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        start = end = head

        # end 조건 들어가야 해서 while문 사용
        while n>0 and end:
            end = end.next
            n -= 1

        # 노드가 하나밖에 없으면 n이 1이라서 end가 none이 됨. 그래서 head.next=none이 되어 에러 없이 실행
        if not end:
            head = head.next
        else:
            while end.next:
                start = start.next
                end = end.next
            start.next = start.next.next
        return head
